pkg/nsexec/nsexec_linux.go

```diff
@@ -0,0 +1,265 @@
+package nsexec
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"strings"
+	"syscall"
+
+	"github.com/opencontainers/runc/libcontainer"
+	"github.com/opencontainers/runc/libcontainer/configs"
+	"github.com/opencontainers/runc/libcontainer/utils"
+	"github.com/sirupsen/logrus"
+	"github.com/vishvananda/netlink/nl"
+)
+
+// Package nsexec is to join container namespace to do some update in container such as update hostname, etc.
+// Calls NsExec with param NsExecOp will do function in container namespace which has been registered
+// by NsExecOp.OpCode. As to realize it, we set a new command nsexec which imports package
+// github.com/opencontainers/runc/libcontainer/nsenter in which use C code to join namespace.
+
+var (
+	regigsterFuncMap = map[string]func(data []byte) (res []byte, err error){}
+)
+
+// Op is Nsexec input, opCode should be registered
+type Op struct {
+	OpCode string `json:"opCode"`
+	Data   []byte `json:"data"`
+}
+
+// Result is the Nsexec result output
+type Result struct {
+	Finish bool   `json:"finish"`
+	Data   []byte `json:"data"`
+	ErrMsg string `json:"errMsg"`
+}
+
+type pid struct {
+	Pid      int `json:"pid"`
+	PidFirst int `json:"pid_first"`
+}
+
+type process struct {
+	initArgs   []string
+	stdin      io.Reader
+	stdout     io.Writer
+	stderr     io.Writer
+	childPipe  *os.File
+	parentPipe *os.File
+	cmd        *exec.Cmd
+}
+
+const stdioFdCount = 3
+
+// Register function which can execute in container namesapce
+func Register(op string, handle func(data []byte) (res []byte, err error)) {
+	if _, exist := regigsterFuncMap[op]; exist {
+		panic(fmt.Sprintf("nsexec func already registered under name %q", op))
+	}
+
+	regigsterFuncMap[op] = handle
+}
+
+func newExecProcess(stdin io.Reader, stdout io.Writer, stderr io.Writer) *process {
+	return &process{
+		initArgs: []string{"/proc/self/exe", "nsexec"},
+		stdin:    stdin,
+		stdout:   stdout,
+		stderr:   stderr,
+	}
+}
+
+func newNsExecCmd(p *process) {
```

Comment:
can we add test cases for this package?

Reply:
The main function NsExec calls a cli command(/proc/self/exe nsexec), it is difficult to add go test for it! We may only test it by cli test.